src/task.c

```diff
@@ -742,6 +742,32 @@ task_perform_one(task_system *ts)
    return rc;
 }
 
+/*
+ * Validate that the task system configuration is basically supportable.
+ * This check protects us from configuring some bizillion #s of background
+ * threads, which somehow sneak through other config parsing checks.
+ * We also enforce other, somewhat arbitrary, limits to leave some headroom
+ * for user-threads to exeecute.
+ */
+static platform_status
+task_config_valid(uint8 num_bg_threads[NUM_TASK_TYPES])
+{
+   uint64 normal_bg_threads   = num_bg_threads[TASK_TYPE_NORMAL];
+   uint64 memtable_bg_threads = num_bg_threads[TASK_TYPE_MEMTABLE];
+
+   if ((normal_bg_threads + memtable_bg_threads) > (MAX_THREADS / 2)) {
+      platform_error_log("Background thread configuration is not supported."
+                         " Total number of background threads configured"
+                         ", normal-bg-threads=%lu, memtable-bg-threads=%lu,"
+                         " must be <= %d\n",
+                         normal_bg_threads,
+                         memtable_bg_threads,
+                         (MAX_THREADS / 2));
+      return STATUS_BAD_PARAM;
```

Comment:
I'm not lovin' the arbitrary limit.  Also, why not check that the thread counts are either both 0 or both non-zero?

Reply:
The both are 0 or both are non-zero is already checked in the test-parsing, `config_parse()`  in config.c (See around L359.)

I'm just trying to create some defenses and guard rails to prevent bad usages being attempted or exercised.
I was just trying to provide a mechanism to protect users from bad usages creeping.

This function exists to check cases where someone does `-num-normal-bg-threads 100000`, or 1 million; or some big big number.

The "<= (MAX_THREADS/2)" is arbitrary ... but I just wanted to implement some checks for preventing large bg-thread counts being configured.

I could change this to something on the lines of:

- check that each bg-thread config param is <= MAX_THREADS
- check that sum of both params is <= MAX_THREADS

The problem I see is that even when (sum of two configs) == MAX_THREADS, and we configure those # of bg-threads ... what do you do next when the 1st end-user threads is created? Things won't work at all.

Let me know what you see as being better, and we can revisit ...

Reply:
> what do you do next when the 1st end-user thread(s) is (are) created? Things won't work at all.

Could this problem instead be addressed by having a check at thread-registration time?

Reply:
We could, ... and maybe the code already does something like that already. Upon inspection, I think what will happen is that we will end up in this assert, which is only an `debug_assert()`.

```
 62 static void
 63 task_init_threadid(task_system *ts)
 64 {

 88 out:
 89    debug_assert(tid != INVALID_TID);
```

I have to verify this with a test-case as to what will happen if someone has consumed all the 63 bits for bg-threads and then a user-thread attempts to register itself.

I feel that kind of lapsed-case scenario (when someone configures all bg-threads) is something we should try to avoid, if possible. Which is what this arbitrary check in this new function is trying to do.

Reply:
I think it's good to error if we know they will definitely hit an assertion later.  But if not, then let them do it, and then let them see an error on thread_register.  This arbitrary limit seems....arbitrary.

Reply:
Change this fn to return an error if sum of # of bg-threads > max-#-of threads.

For the user-thread case, raise & report an error if they have already exceeded bg threads and so registering a new user thread will fail.

Reply:
@rosenhouse , @rtjohnso , @ajhconway - FYI; the bulk of the comments were here. All of these have been addressed and the change set accordingly amended in this rev.